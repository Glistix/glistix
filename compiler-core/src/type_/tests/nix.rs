use crate::{
    glistix_assert_nix_error, glistix_assert_nix_module_error, glistix_assert_nix_module_infer,
};

#[test]
fn nix_unsafe_int_decimal() {
    glistix_assert_nix_error!(
        r#"
  [
    9_223_372_036_854_775_806,
    9_223_372_036_854_775_807,
    9_223_372_036_854_775_808,
    -9_223_372_036_854_775_806,
    -9_223_372_036_854_775_807,
    -9_223_372_036_854_775_808,
  ]
"#
    );
}

#[test]
fn nix_unsafe_int_binary() {
    glistix_assert_nix_error!(
        r#"
  [
    0b111111111111111111111111111111111111111111111111111111111111110,
    0b111111111111111111111111111111111111111111111111111111111111111,
    0b1000000000000000000000000000000000000000000000000000000000000000,
  ]
"#
    );
}

#[test]
fn nix_unsafe_int_octal() {
    glistix_assert_nix_error!(
        r#"
  [
    0o777777777777777777776,
    0o777777777777777777777,
    0o1000000000000000000000,
  ]
"#
    );
}

#[test]
fn nix_unsafe_int_hex() {
    glistix_assert_nix_error!(
        r#"
  [
    0x7FFFFFFFFFFFFFFE,
    0x7FFFFFFFFFFFFFFF,
    0x8000000000000000,
  ]
"#
    );
}

#[test]
fn nix_unsafe_int_in_tuple() {
    glistix_assert_nix_error!(
        r#"
  #(9_223_372_036_854_775_808)
"#
    );
}

#[test]
fn nix_unsafe_int_segment_in_bit_array() {
    glistix_assert_nix_error!(
        r#"
  <<9_223_372_036_854_775_808:64>>
"#
    );
}

#[test]
fn nix_unsafe_int_segment_size_in_bit_array() {
    glistix_assert_nix_error!(
        r#"
  [
    <<0:9_223_372_036_854_775_808>>,
    <<0:size(9_223_372_036_854_775_808)>>,
  ]
"#
    );
}

#[test]
fn nix_unsafe_int_in_const() {
    glistix_assert_nix_module_error!(r#"const i = 9_223_372_036_854_775_808"#);
}

#[test]
fn nix_unsafe_int_in_const_tuple() {
    glistix_assert_nix_module_error!(r#"const i = #(9_223_372_036_854_775_808)"#);
}

#[test]
fn nix_unsafe_int_segment_in_const_bit_array() {
    glistix_assert_nix_module_error!(r#"const i = <<9_223_372_036_854_775_808:64>>"#);
}

#[test]
fn nix_unsafe_int_segment_size_in_const_bit_array() {
    glistix_assert_nix_module_error!(
        r#"const ints = [
  <<0:9_223_372_036_854_775_808>>,
  <<0:size(9_223_372_036_854_775_808)>>,
]"#
    );
}

#[test]
fn nix_unsafe_int_in_pattern() {
    glistix_assert_nix_error!(r#"let assert <<9_223_372_036_854_775_808:64>> = <<>>"#);
}

#[test]
fn nix_unsafe_int_segment_size_in_pattern() {
    glistix_assert_nix_error!(r#"let assert <<0:9_223_372_036_854_775_808>> = <<>>"#);
}

#[test]
fn nix_unsafe_int_with_external_implementation() {
    glistix_assert_nix_module_infer!(
        r#"
@external(nix, "./test.mjs", "go")
pub fn go() -> Int {
  9_223_372_036_854_775_808
}
"#,
        vec![("go", "fn() -> Int")]
    );
}

#[test]
fn nix_unsafe_int_segment_in_pattern_with_external_implementation() {
    glistix_assert_nix_module_infer!(
        r#"
@external(nix, "./test.mjs", "go")
pub fn go(b: BitArray) -> BitArray {
  let assert <<0x800000000000000000:64>> = b
}
"#,
        vec![("go", "fn(BitArray) -> BitArray")]
    );
}

#[test]
fn out_of_range_nix_float() {
    glistix_assert_nix_error!(r#"1.8e308"#);
}

#[test]
fn out_of_range_nix_float_in_pattern() {
    glistix_assert_nix_error!(r#"let assert [1.8e308, b] = [x, y]"#);
}

#[test]
fn out_of_range_nix_float_in_const() {
    glistix_assert_nix_module_error!(r#"const x = 1.8e308"#);
}

#[test]
fn negative_out_of_range_nix_float() {
    glistix_assert_nix_error!(r#"-1.8e308"#);
}

#[test]
fn negative_out_of_range_nix_float_in_pattern() {
    glistix_assert_nix_error!(r#"let assert [-1.8e308, b] = [x, y]"#);
}

#[test]
fn negative_out_of_range_nix_float_in_const() {
    glistix_assert_nix_module_error!(r#"const x = -1.8e308"#);
}
