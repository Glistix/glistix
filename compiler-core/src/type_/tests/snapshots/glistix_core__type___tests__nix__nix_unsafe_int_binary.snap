---
source: compiler-core/src/type_/tests/nix.rs
expression: "\n  [\n    0b111111111111111111111111111111111111111111111111111111111111110,\n    0b111111111111111111111111111111111111111111111111111111111111111,\n    0b1000000000000000000000000000000000000000000000000000000000000000,\n  ]\n"
---
----- SOURCE CODE

  [
    0b111111111111111111111111111111111111111111111111111111111111110,
    0b111111111111111111111111111111111111111111111111111111111111111,
    0b1000000000000000000000000000000000000000000000000000000000000000,
  ]


----- ERROR
error: Int is outside Nix's integer range
  ┌─ /src/one/two.gleam:5:5
  │
5 │     0b1000000000000000000000000000000000000000000000000000000000000000,
  │     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

This integer value is too large to be represented by Nix's integer type. To
avoid this error integer values must be in the range -(2^63 - 1) - (2^63 -
1).
